lib/spack/spack/install_test.py

```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
     return names[0]
 
 
+def package_class(spec_or_pkg):
+    """Return the package class for the spec or package.
+
+    Args:
+        spec_or_pkg (spack.spec.Spec or spack.package_base.PackageBase): spec or
+            package (class)
+
+    Raises: ValueError: If not given a spec, package, or class
+    """
+    if isinstance(spec_or_pkg, spack.spec.Spec):
+        spec = spec_or_pkg
+        try:
+            cls = spec.package.__class__
+        except AssertionError:
+            try:
+                cls = spec.package_class
+            except spack.repo.UnknownPackageError as e:
+                tty.debug("{0}: Cannot retrieve class from spec: {1}".format(spec.name, str(e)))
+                return None
+
+    elif isinstance(spec_or_pkg, spack.package_base.PackageBase):
+        cls = spec_or_pkg.__class__
+
+    elif inspect.isclass(spec_or_pkg):
+        cls = spec_or_pkg
+
+    else:
+        raise ValueError("Cannot retrieve package class for {0}".format(spec_or_pkg))
+
+    return cls
+
+
+def test_functions(spec_or_pkg, add_virtuals=False, names=False):
+    """Grab all non-pass-only test functions (names) associated with the spec or package.
+
+    Args:
+        spec_or_pkg (spack.spec.Spec or spack.package_base.PackageBase): spec or
+            package (class)
+        add_virtuals (bool): True will add any test methods of provided virtuals
+            when a package is provided
+        names (bool): True results in the return of test function names
+
+    Returns: list of test functions or function names
+
+    Raises: ValueError: If not given a spec, package, or class
+    """
+    if add_virtuals and isinstance(spec_or_pkg, spack.package_base.PackageBase):
+        pkg = spec_or_pkg
+        v_names = virtuals(pkg)
+        test_specs = [pkg.spec] + [spack.spec.Spec(v_name) for v_name in sorted(v_names)]
+        methods = []
+        for spec in test_specs:
+            methods.extend(test_functions(spec, add_virtuals=False, names=names))
+        return methods
+
+    cls = package_class(spec_or_pkg)
+    if not cls:
```

Comment:
(Minor) I think this might be slightly safer for future refactoring:
```suggestion
    if cls is None:
```